### ISSUE TITLE
Reference cycle in _LRScheduler

### ISSUE THREAD

**Issue by huzecong** [pytorch/pytorch#25605]
## To Reproduce

In the following code introduced in version 1.2.0 by #20124, when constructing the `_LRScheduler`, the `step` method of `optimizer` is replaced with a new closure that captures the `optimizer` itself.

https://github.com/pytorch/pytorch/blob/9cb9f15989b08d71fcf0ad0ddda84299c0854c2d/torch/optim/lr_scheduler.py#L29-L40

This makes the optimizer reference itself, thus creating a reference cycle that prevents it from being garbage collected.

To verify this, run the following snippet:
```python
import torch
from torch import nn

import gc

def main():
    param = nn.Parameter(torch.randn(10))

    optim = torch.optim.Adam([param])
    scheduler = torch.optim.lr_scheduler.LambdaLR(optim, lambda epoch: 1.0)
    del scheduler

    print(gc.get_referrers(optim))
    
    gc.collect()
    del optim
    print(gc.collect())

if __name__ == '__main__':
    main()
```

The output is:
```
[<bound method Adam.step of Adam (
Parameter Group 0
    amsgrad: False
    betas: (0.9, 0.999)
    eps: 1e-08
    initial_lr: 0.001
    lr: 0.001
    weight_decay: 0
)>, <cell at 0x109354948: Adam object at 0x10940d8d0>]
12
```
which shows that `optim.step` is a referrer of `optim`, and it's not collected automatically.

If the `scheduler` is not created, the output becomes:
```
[]
0
```
which shows that no referrers of `optim` exist and it's GC-ed properly.

## Expected behavior

The self-reference could be easily prevented by using the `self` argument and wrapping the created function with `types.MethodType(wrapper, opt)`.

## Environment

 - PyTorch Version: 1.2.0
 - Other information probably irrelevant.

cc @ezyang @gchanan @zou3519 @vincentqb

**Comment by soumith** [comment]
this is probably high-pri because holding on to memory is not good for training.
Also marking it for next release

**Comment by ezyang** [comment]
@huzecong This is a nice catch. Do you think you could submit a Pr for this?

**Comment by huzecong** [comment]
@ezyang Sure, I can do that. But I'm wondering what would be the ideal implementation? Comments in #20124 suggest that the current code for detecting calls to `optim.step` may be missing a few cases.

**Comment by ezyang** [comment]
It's probably better to fix the reference cycle first, and handle the comment on the PR separately. The comment does seem like a reasonable thing to also fix, so if you want to roll them together in one PR, that'd be fine by me too.

**Comment by huzecong** [comment]
I think I'll fix the cycle & the problem mentioned in that issue thread together. Should probably take a day or two.

**Comment by huzecong** [comment]
PR #25776 submitted to resolve this issue. Note that this only included the fix for cyclic references.

It's actually quite complicated to solve other related issues. For instance, code for issuing the warning is in `step()` method of the base class, while certain subclasses (`ReduceLROnPlateau` and `CosineAnnealingWarmRestarts`) override `step()` without calling `super()`.

I propose we change the interface of LRScheduler and split `step()` into three methods: `_update_last_epoch()`, `_update_lr()`, and `step()`.

- `step()` internally calls `_update_last_epoch()`, checks whether a warning should be issues, and then calls `_update_lr()`.
- `_update_last_epoch()` updates `self.last_epoch` based on the `epoch` argument pass to `step()`. For most schedulers this would just be:
  ```python
  def _update_last_epoch(epoch=None):
      if epoch is None:
          epoch = self.last_epoch + 1
      self.last_epoch = epoch
  ```
  but `ReduceLROnPlateau` and `CosineAnnealingWarmRestarts` have their unique implementations.
- `_update_lr()` calls `self.get_lr()` and updates `param_group["lr"]`. For most schedulers this would just be:
  ```python
  for param_group, lr in zip(self.optimizer.param_groups, self.get_lr()):
      param_group['lr'] = lr
  ```
  while for `ReduceLROnPlateau` this would be similar to the current `_reduce_lr()` method. It might also be beneficial to move the logic for computing new LR values in `ReduceLROnPlateau._reduce_lr()` to `get_lr()` to be consistent with other classes.

I would love to hear your opinions on this.

**Comment by ezyang** [comment]
^-- @vincentqb

**Comment by vincentqb** [comment]
> ^-- @vincentqb

in the middle of typing ;) thanks for the ping :)

**Comment by vincentqb** [comment]
Thanks for the work on the PR and the feedback! And thanks for catching the garbage collection issue!

Your proposal raises interesting points. :) Can you elaborate on how the proposal to change the interface relates to fixing the garbage collection? or catching when the scheduler is not called at the right time? Do you think your interface suggestion simplifies the scheduler code independently of fixing the garbage collection?

I would really like if the scheduler code is as "simple"/"readable" as possible so other users can feel empowered to modify it and tailor it to their needs. In fact, I'm hoping with time, we will be able to remove that logic to catch the stepping order completely as users will be familiar with the new documented interface. :)

By the way, some changes (e.g. [here](https://github.com/pytorch/pytorch/pull/24352/files#diff-5fc1045ac3d323f96b6987e78fb1e468R128)) in #24352 would relate to your interface suggestions.

**Comment by huzecong** [comment]
@vincentqb Thanks for the comments! My proposal here is not directly related to cyclic references. It's trying to address the issue that subclasses that override `step()` would have to copy-paste the logic for detecting "scheduler-called-before-optimizer" that's defined in `_LRScheduler.step()`. I agree that this might hurt readability, and it might be better (and simpler) to just extract the detection part into a separate method.

**Comment by vincentqb** [comment]
By the way, [this](https://github.com/pytorch/pytorch/issues/17630#issuecomment-525928944) in #17630 and [this](https://github.com/pytorch/pytorch/issues/9942#issuecomment-520638174) in #9942 seems related to this PR. Thoughts?
